consul_exporter.go

```diff
@@ -0,0 +1,216 @@
+package main
+
+import (
+	"flag"
+	_ "fmt"
+	_ "io"
+	_ "io/ioutil"
+	"log"
+	_ "net"
+	"net/http"
+	_ "strconv"
+	_ "strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace = "consul"
+)
+
+var (
+	serviceLabelNames = []string{"service", "node"}
+	memberLabelNames  = []string{"member"}
+)
+
+// Exporter collects HAProxy stats from the given URI and exports them using
+// the prometheus metrics package.
+type Exporter struct {
+	URI   string
+	mutex sync.RWMutex
+
+	up, clusterServers              prometheus.Gauge
+	totalQueries, jsonParseFailures prometheus.Counter
+	serviceMetrics, lockMetrics     map[string]*prometheus.GaugeVec
+	client                          *api.Client
+}
+
+func newServiceMetric(metricName string, docString string, constLabels prometheus.Labels) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace:   namespace,
+			Name:        "service_" + metricName,
+			Help:        docString,
+			ConstLabels: constLabels,
+		},
+		serviceLabelNames,
+	)
+}
+
+// NewExporter returns an initialized Exporter.
+func NewExporter(uri string, consulLocks string, timeout time.Duration) *Exporter {
+	// connect to Consul
+
+	consul_client, _ := api.NewClient(&api.Config{
+		Address: uri,
+	})
+
+	// init our exporter
+
+	return &Exporter{
+		URI: uri,
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "up",
+			Help:      "Was the last query of Consul successful.",
+		}),
+		totalQueries: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_total_queries",
+			Help:      "Current total Consul queries.",
+		}),
+
+		clusterServers: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "cluster_servers",
+			Help:      "How many peers are in the cluster.",
+		}),
+
+		serviceMetrics: map[string]*prometheus.GaugeVec{
+			"nodes_healthy":   newServiceMetric("nodes", "Number of nodes", prometheus.Labels{"healthy": "healthy"}),
+			"nodes_unhealthy": newServiceMetric("nodes", "Number of nodes", prometheus.Labels{"healthy": "unhealthy"}),
+		},
+
+		client: consul_client,
+	}
+}
+
+// Describe describes all the metrics ever exported by the HAProxy exporter. It
+// implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	ch <- e.up.Desc()
+	ch <- e.totalQueries.Desc()
+	ch <- e.clusterServers.Desc()
+}
+
+// Collect fetches the stats from configured HAProxy location and delivers them
+// as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	services := make(chan *api.ServiceEntry)
+
+	go e.queryClient(services)
+
+	e.mutex.Lock() // To protect metrics from concurrent collects.
+	defer e.mutex.Unlock()
+
+	// reset metrics
```

Comment:
Nitpick: we use sentence-style for comments everywhere, a la "Reset metrics.". Especially for multi-line comments like one of the comments you have below, that improves readability/cleanliness.